kolibri_app_desktop_xdg_plugin/channel_launchers.py

```diff
@@ -34,24 +36,59 @@
 
 
 def update_channel_launchers(force=False):
-    launchers_from_db = list(ChannelLauncher_FromDatabase.load_all())
-    launchers_from_disk = list(ChannelLauncher_FromDisk.load_all())
+    context = ChannelLaunchersContext()
+
+    launchers_from_db = list(ChannelLauncher_FromDatabase.load_all(context=context))
+    launchers_from_disk = list(ChannelLauncher_FromDisk.load_all(context=context))
+    did_icons_change = False
 
     for launcher in launchers_from_disk:
         if not any(map(launcher.is_same_channel, launchers_from_db)):
             logger.info("Removing desktop launcher %s", launcher)
             launcher.delete()
+            did_icons_change = True
 
     for launcher in launchers_from_db:
         if not any(map(launcher.is_same_channel, launchers_from_disk)):
             logger.info("Creating desktop launcher %s", launcher)
             launcher.save()
+            did_icons_change = True
         elif force or any(map(launcher.compare, launchers_from_disk)):
             logger.info("Updating desktop launcher %s", launcher)
             launcher.save()
+            did_icons_change = True
+
+    if did_icons_change:
+        try:
+            system_theme_index = "/usr/share/icons/hicolor/index.theme"
+            theme_index = os.path.join(context.icon_theme_dir, "index.theme")
+            shutil.copyfile(system_theme_index, theme_index)
+        except OSError:
+            subprocess.run(["gtk-update-icon-cache", context.icon_theme_dir])
+        else:
+            subprocess.run(
+                [
+                    "gtk-update-icon-cache",
+                    "--ignore-theme-index",
+                    context.icon_theme_dir,
+                ]
+            )
+
+
+class ChannelLaunchersContext(object):
+    @property
+    def applications_dir(self):
+        return os.path.join(get_content_share_dir_path(), "applications")
+
+    @property
+    def icon_theme_dir(self):
+        return os.path.join(get_content_share_dir_path(), "icons", "hicolor")
 
 
 class ChannelLauncher(object):
+    def __init__(self, context):
```

Comment:
The context here is the first positional argument but in the `load_all` call you're passing the `context` named argument, so if I'm understanding correctly, this context will be always `None` when calling the `load_all` in `ChannelLauncher_FromDisk` and `ChannelLauncher_fromDatabase`.

I think that this should be `context=None` or if it's a positional argument, call it like `load_all(context)`

Reply:
Interestingly, Python does sort that out (you can treat a positional argument as a named argument), but this is indeed _really_ unclear. I changed this to be more explicit, replacing `*args, **kwargs` with just `context`, since that flexibility isn't useful here anyway :) Thanks for pointing it out!